MNTOR-1878: API for location autocomplete

flozia opened the pull request:
<!-- The following is intended to be helpful to you. Feel free to remove anything that is not. -->

# References: 
Jira: [MNTOR-1878](https://mozilla-hub.atlassian.net/browse/MNTOR-1878)
Figma: [Enter details inputs](https://www.figma.com/file/CaEKIhvSJqf6KNIMzSkt40/Concepts-for-Monitor-MVP-Redesign?type=design&node-id=1205-61900&mode=design&t=xa4nmpd5bdTjhiM7-0)


<!-- When adding a new feature: -->

# Description

This PR adds the data and API needed for the location auto-completion input during the “Personal info” step for the breaches scan.

**Location data**
In order to create a list of US cities with the features we need we can use the script `createLocationAutocompleteData` that downloads and parses data from [GeoNames](https://www.geonames.org/). In order to get a dataset that supports our use case we compile a JSON file from three exported table dumps: https://www.geonames.org/export/dump/.
- Locations: https://download.geonames.org/export/dump/US.zip
All US location features that we filter to only get populated places.
- Alternate names: https://download.geonames.org/export/dump/alternateNames.zip
Alternate location names that we can use to make searching for locations more user-friendly.
- Hierarchy: https://download.geonames.org/export/dump/hierarchy.zip
Used to filter out locations that have populated places as a parent and thus are not cities but districts.

**Location autocomplete API**
The API `location-autocomplete` takes a search query and provides location suggestions. With https://github.com/leeoniya/uFuzzy we search through the list of cities that we created with the script `createLocationAutocompleteData`.


# Screenshot

https://github.com/mozilla/blurts-server/assets/13835474/343fba64-443d-4af1-9ffd-fd6a7872b297

# How to test

1. Before running the site install the newly added dependencies.
```
npm install
```

2. Create location data: Running the script takes a few minutes. While running a temporary directory `tmp-data` is created and deleted after the script ran successfully. The resulting `locationAutocompleteData.json` file will be put in the root directory.
```
npx ts-node-esm src/scripts/build/createLocationAutocompleteData/index.ts
```

3. You can test the location autocompletion API using the test input on http://localhost:6060/admin/location-search. Make sure you are signed in and added yourself as an admin in the `.env` file under `ADMINS=`.

Vinnl commented on package.json:
```diff
@@ -49,11 +49,11 @@
     "@fluent/bundle": "^0.17.1",
     "@fluent/langneg": "^0.6.2",
     "@fluent/react": "^0.15.0",
+    "@leeoniya/ufuzzy": "^1.0.8",
```
Just wanted to call out that this is a relatively new ([9 months old](https://github.com/leeoniya/uFuzzy/commit/f920effc32934fd014a0f8b2d9a5533504d6a383)) and not-that-widely-used library with a single maintainer. It's also probably relatively low-risk and [not particularly big](https://bundlephobia.com/package/@leeoniya/ufuzzy@1.0.8), but calling out in case I missed a consideration.

flozia replied:
There are definitely a couple of libraries that we could use here. This is the one I found gave the best results for the location data we have without diving too deep into their customization options. I gave a couple a quick try without comprehensive documentation:

- FlexSearch: https://github.com/nextapps-de/flexsearch
- Fuse.js: https://fusejs.io/
- Lunr.js: https://github.com/olivernn/lunr.js
- match-sorter: https://github.com/kentcdodds/match-sorter

Happy to try and make one of the previous options or another one work for us.

Vinnl commented on src/scripts/createLocationAutocompleteData/types.d.ts:
Could you move these types into the regular `.ts` file? `.d.ts` files are [generally for typing external libraries](https://www.typescriptlang.org/docs/handbook/modules.html#working-with-other-javascript-libraries) that do not come with their own type definitions (or that we're monkeypatching), which we hopefully won't have to do as much anymore now that we're on React and proper TypeScript.

Nice job on the extensive types though! :clap:

flozia replied:
Changed [the file type](https://github.com/mozilla/blurts-server/pull/3186/files#diff-49b0d5780cc466856188c6ecc30300fc5dccaa91643362fccee8e24d5c3d5b7d).

Vinnl commented on src/scripts/createLocationAutocompleteData/types.d.ts:
```diff
@@ -0,0 +1,153 @@
+ * https://download.geonames.org/export/dump/.
+ */
+
+type TAlternateNameId = string; // the id of this alternate name
```
If you use JSDoc strings (`/** ... */`), these descriptions will be visible when someone hovers something of this type.

flozia replied:
Great suggestion. Updated the comments: [c2b8a7a](https://github.com/mozilla/blurts-server/pull/3186/commits/c2b8a7aea10d32af606a0770108dac44acd6fe6d).

Vinnl commented on src/scripts/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,345 @@
+  writeStream,
+}: {
+  url: string;
+  writeStream: any;
```
I think this type can be imported from `fs` as well:

```suggestion
  writeStream: WriteStream;
```

Vinnl commented on src/scripts/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,345 @@
+  const startTime = Date.now();
+  console.info("Create autocomplete location data");
+
+  const __dirname = dirname(fileURLToPath(import.meta.url));
+  const tmpDirPath = resolve(__dirname, "tpm-data");
```
You might want to use [`os.tmpdir`](https://nodejs.org/api/os.html#ostmpdir) here instead.

flozia replied:
Good idea: [2a95570](https://github.com/mozilla/blurts-server/pull/3186/commits/2a955703d3459c925574942f708b9e853d9f4643).

Vinnl commented on src/scripts/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,345 @@
+    (locationPopulated, rowIndex) => {
+      logProgress(rowIndex, locationDataPopulatedCount);
+
+      const hasPopulatedParentLocation = hierarchyIds.some(
+        ([parentId, childId]) => {
+          if (locationPopulated.id === childId) {
+            return locationDataPopulated.some((location) => {
+              if (
+                location.id === parentId &&
+                location.featureClass === allowedFeatureClass
+              ) {
+                return true;
+              }
+              return false;
+            });
+          }
+          return false;
+        }
```
Not necessary at all, but you can shave off some indenting if you put the bailout conditions first (no guarantee that I didn't make mistakes below, typing in the GitHub text box):

```suggestion
      const hasPopulatedParentLocation = hierarchyIds.some(
        ([parentId, childId]) => {
          if (locationPopulated.id !== childId) {
            return false;
          }
          return locationDataPopulated.some((location) => {
            return (
              location.id === parentId &&
              location.featureClass === allowedFeatureClass
            );
          });
          
        }
```

flozia replied:
That does look better: [9d94401](https://github.com/mozilla/blurts-server/pull/3186/commits/9d94401ff628462e4795246d0dc64dc14d86931e).

Vinnl commented on src/scripts/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,345 @@
+
+  console.info(`Writing location data to file: ${LOCATIONS_DATA_FILE}`);
+  const locationDataFinal = {
+    name: "monitor-location-autocomplete-data",
```
Nice job storing this metadata as well :clap:

Vinnl commented on src/scripts/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,345 @@
+  console.error("Creating location file failed with:", error);
+}
+
+process.exit();
```
Is this needed?

flozia replied:
Removed.

Vinnl commented on src/scripts/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,345 @@
+const FETCH_REMOTE_DATASETS = true;
+const CLEANUP_TMP_DATA_AFTER_FINISHED = true;
+
+// Only include populated place a city, town, village, or other
+// agglomeration of buildings where people live and work.
```
I think a word or character (`:`?) is missing here - I don't quite know what this is supposed to say :sweat_smile: 

(Alternatively, I should drink some coffee.)

Vinnl commented on src/app/api/v1/location-autocomplete/route.ts:
```diff
@@ -0,0 +1,114 @@
+  );
+
+  const resultsOrdered = order.map((orderIndex) => results[orderIndex]);
+  // Split and only sort the first x percentage of results by population.
```
Could you expand on this comment why we're doing this? My interpretation is that this takes the top 75% most relevant results, but then ranks the ones with the largest population first - wouldn't that potentially lead to the closest match only being 75% down the list?

flozia replied:
That’s right. The idea is to improve the results and take the population into account to improve the “scores” for matches, but not including very weak matches. This especially improves the suggestions for short search queries that can have a lot of matches.

Expanded on the comment: [339d004](https://github.com/mozilla/blurts-server/pull/3186/commits/339d0043408d90becb51514969ea1113a17fb916).

Vinnl commented on src/app/components/client/LocationAutocompleteInput.tsx:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    return () => {
+      abortController.abort();
```
Just verifying that this appears to work well; if a new request is sent while the previous response hasn't come in yet, it'll abort the request.

![image](https://github.com/mozilla/blurts-server/assets/4251/45b89645-e19a-4d78-9821-4527a1547e9c)

Vinnl reviewed (commented):
Instead of checking in `locationAutocompleteData.json` directly, would it make sense to run `createLocationAutocompleteData` in the `npm run build` step, and add it to the `.gitignore`? And then maybe we should move it into a `build/` folder, to distinguish it from cron job scripts? (Especially since we'll probably want to move cron job scripts to a separate project later (MPP-1880).)

Other than that, the main thing is to not use a `.d.ts` file. The rest appears to work well, though I'll leave it to actual back-enders to approve :)

flozia commented:
> Instead of checking in `locationAutocompleteData.json` directly, would it make sense to run `createLocationAutocompleteData` in the `npm run build` step, and add it to the `.gitignore`? And then maybe we should move it into a `build/` folder, to distinguish it from cron job scripts? (Especially since we'll probably want to move cron job scripts to a separate project later (MPP-1880).)

I had it in the `.gitignore` before and only added `locationAutocompleteData.json` later in https://github.com/mozilla/blurts-server/pull/3186/commits/1ce38b024ee65cd34d41e112fa94e388dcb54315.  I was not sure if adding it to the build script increases the time by too much it takes to build the site, but thinking about it again I guess this is also a good way to keep it up-to-date. Added the script to the build step and moved it into a separate folder in [c89769d](https://github.com/mozilla/blurts-server/pull/3186/commits/c89769d855b58c0ebad8424b6f5ca701a4c0e6a6) as suggested.

I’ll address the other review comments inline.

> The rest appears to work well, though I'll leave it to actual back-enders to approve :)

Thank you for the review and your suggestions @Vinnl! :)

pdehaan commented on src/app/api/v1/location-autocomplete/route.ts:
```diff
@@ -0,0 +1,119 @@
+
+export type TMatchingLocations = Array<IRelevantLocation> | [];
+
+export interface ISearchLocationParams {
+  searchQuery: string;
+  config: {
+    minQueryLength: number;
+    maxResults: number;
+  };
+}
+
+export interface ISearchLocationResults {
+  searchQuery: string;
+  results: TMatchingLocations;
+}
```
Related to above comment about interfaces being prefixed w/ `I`:

<details>
<summary><kbd>git grep -Pho "(export )?interface \w+" | sort | uniq</summary>

```sh
export interface Breach
export interface BreachResolution
export interface BreachResolutionRequest
export interface BreachStats
export interface CircleChartProps
export interface IDataFileUrls
export interface IRelevantLocation
export interface IRelevantLocationAlternate
export interface ISearchLocationParams
export interface ISearchLocationResults
export interface Subscriber
export interface SubscriberBreachResolution
export interface SubscriberEmail
export interface UserBreaches
export interface VerifiedEmail
interface BreachRow
interface EmailAddRequest
interface EmailAddressRow
interface EmailDeleteRequest
interface EmailResendRequest
interface EmailUpdateCommOptionRequest
interface FeatureFlagRow
interface FxaProfile
interface IntrinsicElements
interface JWT
interface JwtPayload
interface PasswordChangeEvent
interface Profile
interface ProfileChangeEvent
interface Session
interface SubscriberRow
interface SubscriptionStateChangeEvent
interface Tables
interface Window
```
</details>

```sh
git grep -Pon "(?:export )?interface [I][A-Z]\w+" -- '*.ts' '*.tsx' | sort | uniq

src/app/api/v1/location-autocomplete/route.ts:16:export interface ISearchLocationParams
src/app/api/v1/location-autocomplete/route.ts:24:export interface ISearchLocationResults
src/scripts/build/createLocationAutocompleteData/types.ts:846:export interface IRelevantLocation
src/scripts/build/createLocationAutocompleteData/types.ts:857:export interface IRelevantLocationAlternate
src/scripts/build/createLocationAutocompleteData/types.ts:867:export interface IDataFileUrls
```

flozia replied:
Also removed in [e652217](https://github.com/mozilla/blurts-server/pull/3186/commits/e652217a7d9eca0e6b324c2fdca92e6d7dddf49a).

pdehaan commented on src/app/api/v1/location-autocomplete/route.ts:
```diff
@@ -0,0 +1,119 @@
+// eslint-disable-next-line import/no-unresolved
+import locationData from "../../../../../locationAutocompleteData.json";
+
+export type TMatchingLocations = Array<IRelevantLocation> | [];
```
What's `TMatchingLocations`'s `T` prefix, does that stand for `Type`?
Seems like we don't regularly use that pattern for types:

<details>
<summary><kbd>git grep -Poh "(?:export )?type \w+" -- '*.ts' '*.tsx' | sort | uniq</kbd></summary>

```sh
export type BreachResolutionTypes
export type ExtendedReactLocalization
export type FeatureFlag
export type FeatureFlagPutRequest
export type GetFragment
export type HibpBreachDataTypes
export type LocaleData
export type Props
export type TAlternateNameData
export type TLocationData
export type TMatchingLocations
type AppPickerItemProps
type AppPickerMenuProps
type AppPickerPopoverProps
type AppPickerTriggerProps
type BreachAutoInsertedColumns
type BreachOptionalColumns
type DetailsFoundProps
type EmailAddressAutoInsertedColumns
type FeatureFlagAutoInsertedColumns
type FeatureFlagOptionalColumns
type FxaProfile
type FxaSubscriptionResponse
type LogFunction
type Options
type Story
type SubscriberAutoInsertedColumns
type SubscriberOptionalColumns
type TAlternateName
type TAlternateNameId
type TFrom
type TGeonameId
type TIsAbbreviation
type TIsColloquial
type TIsHistoric
type TIsPreferredName
type TIsShortName
type TIsolanguage
type TLocationAdmin1Code
type TLocationAdmin2Code
type TLocationAdmin3Code
type TLocationAdmin4Code
type TLocationAlternatenames
type TLocationAsciiname
type TLocationCc2
type TLocationCountryCode
type TLocationDem
type TLocationElevation
type TLocationFeatureClass
type TLocationFeatureCode
type TLocationLatitude
type TLocationLongitude
type TLocationModificationDate
type TLocationName
type TLocationPopulation
type TLocationTimezone
type TTo
type assertion
type checking
```
</details>

In fact, if we filter the results to T* prefixes it looks like we only use that pattern in 1-2 files added here:

<details>
<summary><kbd>git grep -Pon "(?:export )?type \w+" -- '*.ts' '*.tsx' | sort | uniq | grep "\btype\sT"</kbd></summary>

```sh
src/app/api/v1/location-autocomplete/route.ts:14:export type TMatchingLocations
src/scripts/build/createLocationAutocompleteData/types.ts:104:type TLocationFeatureCode
src/scripts/build/createLocationAutocompleteData/types.ts:17:type TAlternateNameId
src/scripts/build/createLocationAutocompleteData/types.ts:19:type TGeonameId
src/scripts/build/createLocationAutocompleteData/types.ts:36:type TIsolanguage
src/scripts/build/createLocationAutocompleteData/types.ts:38:type TAlternateName
src/scripts/build/createLocationAutocompleteData/types.ts:40:type TIsAbbreviation
src/scripts/build/createLocationAutocompleteData/types.ts:42:type TIsPreferredName
src/scripts/build/createLocationAutocompleteData/types.ts:44:type TIsShortName
src/scripts/build/createLocationAutocompleteData/types.ts:49:type TIsColloquial
src/scripts/build/createLocationAutocompleteData/types.ts:54:type TIsHistoric
src/scripts/build/createLocationAutocompleteData/types.ts:56:type TFrom
src/scripts/build/createLocationAutocompleteData/types.ts:58:type TTo
src/scripts/build/createLocationAutocompleteData/types.ts:60:export type TAlternateNameData
src/scripts/build/createLocationAutocompleteData/types.ts:74:type TLocationName
src/scripts/build/createLocationAutocompleteData/types.ts:76:type TLocationAsciiname
src/scripts/build/createLocationAutocompleteData/types.ts:787:type TLocationCountryCode
src/scripts/build/createLocationAutocompleteData/types.ts:792:type TLocationCc2
src/scripts/build/createLocationAutocompleteData/types.ts:797:type TLocationAdmin1Code
src/scripts/build/createLocationAutocompleteData/types.ts:803:type TLocationAdmin2Code
src/scripts/build/createLocationAutocompleteData/types.ts:805:type TLocationAdmin3Code
src/scripts/build/createLocationAutocompleteData/types.ts:807:type TLocationAdmin4Code
src/scripts/build/createLocationAutocompleteData/types.ts:809:type TLocationPopulation
src/scripts/build/createLocationAutocompleteData/types.ts:811:type TLocationElevation
src/scripts/build/createLocationAutocompleteData/types.ts:818:type TLocationDem
src/scripts/build/createLocationAutocompleteData/types.ts:81:type TLocationAlternatenames
src/scripts/build/createLocationAutocompleteData/types.ts:820:type TLocationTimezone
src/scripts/build/createLocationAutocompleteData/types.ts:822:type TLocationModificationDate
src/scripts/build/createLocationAutocompleteData/types.ts:824:export type TLocationData
src/scripts/build/createLocationAutocompleteData/types.ts:83:type TLocationLatitude
src/scripts/build/createLocationAutocompleteData/types.ts:85:type TLocationLongitude
src/scripts/build/createLocationAutocompleteData/types.ts:90:type TLocationFeatureClass
```
</details>

flozia replied:
Yes, that’s supposed to stand for `Type`. Thanks for the callout — that is a pattern I didn’t explicitly plan on introducing. Removed with [e652217](https://github.com/mozilla/blurts-server/pull/3186/commits/e652217a7d9eca0e6b324c2fdca92e6d7dddf49a).

pdehaan commented on src/app/components/client/LocationAutocompleteInput.tsx:
```diff
@@ -0,0 +1,96 @@
+
+  return (
+    <div>
+      <input onInput={handleOnInput} placeholder="Enter city and state" />
```
nit: Place holder should be localized.

flozia replied:
This string will be localized with the PR that implements the actual view this component will be used in. Right now this component is just being used for testing this PR.

rhelmer commented on package.json:
```diff
@@ -9,7 +9,7 @@
   "scripts": {
     "dev": "next dev --port=6060",
-    "build": "next build",
+    "build": "npx ts-node-esm src/scripts/build/createLocationAutocompleteData/index.ts && next build",
```
I don't think we should use `npx` here, we should just install `ts-node-esm` as a dev dependency. It's better to pin and let dependabot et al update vs. not knowing what version was used etc.

pdehaan replied:
`ts-node-esm` is packaged with `ts-node` dependency.

```sh
cat package.json | jq '.devDependencies'

{ "ts-node": "^10.9.1" }
```

```sh
ls -l ./node_modules/.bin

acorn -> ../acorn/bin/acorn
ts-node -> ../ts-node/dist/bin.js
ts-node-cwd -> ../ts-node/dist/bin-cwd.js
ts-node-esm -> ../ts-node/dist/bin-esm.js
ts-node-script -> ../ts-node/dist/bin-script.js
ts-node-transpile-only -> ../ts-node/dist/bin-transpile.js
ts-script -> ../ts-node/dist/bin-script-deprecated.js
tsc -> ../typescript/bin/tsc
tsserver -> ../typescript/bin/tsserver
```

flozia replied:
That’s right! Calling `ts-node-esm` directly now instead of through `npx`: [2c46792](https://github.com/mozilla/blurts-server/pull/3186/commits/2c467928164054628ac090b3bde1e130ffad3e79).

flozia commented on src/scripts/build/createLocationAutocompleteData/index.ts:
```diff
@@ -0,0 +1,353 @@
+  console.info("Create autocomplete location data");
+
+  const __dirname = dirname(fileURLToPath(import.meta.url));
+  const tmpDirPath = resolve(__dirname, "tpm-data");
```
Not sure why the change didn’t show up but I changed the [tmp dir path](https://github.com/mozilla/blurts-server/pull/3186/files#diff-799a42b2e89cbc5107d1274341121528fb3627d8ca50ce76de38e042efa41344R112) in https://github.com/mozilla/blurts-server/pull/3186/commits/2a955703d3459c925574942f708b9e853d9f4643.

flozia commented on src/app/(nextjs_migration)/(authenticated)/admin/location-search/page.tsx:
```diff
@@ -0,0 +1,15 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This page is only meant for testing the implementation and this PR. I’ll remove it as soon as I’ll integrate it with the welcome flow.

flozia commented on package.json:
```diff
@@ -9,7 +9,7 @@
   "scripts": {
     "dev": "next dev --port=6060",
-    "build": "next build",
+    "build": "npx ts-node-esm src/scripts/build/createLocationAutocompleteData/index.ts && next build",
```
Sounds good — moved to a separate command in [2c46792](https://github.com/mozilla/blurts-server/pull/3186/commits/2c467928164054628ac090b3bde1e130ffad3e79).

mansaj reviewed (approved):
Pending fixes from other review comments

flozia commented on src/app/api/v1/location-autocomplete/route.ts:
```diff
@@ -0,0 +1,121 @@
+
+async function getLocationsByQuery(searchQuery: string) {
+  if (!locationData) {
+    console.error("No location data available.");
```
Done in https://github.com/mozilla/blurts-server/pull/3186/commits/ccea89babe067f75a0221b15ca9bf0bf6acc5688.

mansaj commented on package.json:
```diff
@@ -9,7 +9,7 @@
   "scripts": {
     "dev": "next dev --port=6060",
-    "build": "next build",
+    "build": "npm run create-location-data && next build",
```
It will error out, correct. If you go to the temp `admin/` UI, it will throw an error

flozia replied:
It’s also a helpful error message saying:
```
No location data available: You may need to run `npm run create-location-data`.
```

flozia commented:
> * add note to README about it

Added in commit [6ee1ee7](https://github.com/mozilla/blurts-server/pull/3186/commits/6ee1ee7aa5dea1b207dab992378345e6f3e7ccbb).

> * consider filing an issue to move this to a cron job that populates CDN

Yes, I agree that moving this to a cronjob is a good next step: Created [MNTOR-1898](https://mozilla-hub.atlassian.net/browse/MNTOR-1898).
